### PR TITLE
301   profile picture

### PR DESCRIPTION
**Pull request by EduardoLopezPalomo**
When you go to myProfile page there is a button and you can upload an image from your computer. If there is no picture in the database the image will be a default profile icon.
In the member list the profilepicture will be the profile picture that the user upload in his myprofile page.

When you create an event it will also has a button for uploading a image. And when you edit it you can also change the picture

### CODE REVIEW

**Review by jannejjj — COMMENTED** [review]
Left a couple of comments, but in general looks nice! Didn't test yet, will do that a bit later.

File: server/routes/api.js [hunk 1]
```diff
@@ -4,6 +4,8 @@ var router = express.Router();
 const {body, validationResult } = require("express-validator");
 const Member = require("../models/member");
 const Member_event = require("../models/member_event");
+const Event = require("../models/event");
```
Comment (jannejjj) [inline server/routes/api.js:7]: Event was already declared, now it's there twice! :D Remove one of them

File: client/src/modals/EditEventModal.js [hunk 2]
```diff
@@ -106,6 +107,12 @@ function EditEventModal(props)
                   value={props.tickets}
                   onChange={e => props.setTickets(e.target.value)}
                 />
+                <h3>
+                  Upload profile image
```
Comment (jannejjj) [inline client/src/modals/EditEventModal.js:111]: Change to upload "event" image :P

**Review by jannejjj — COMMENTED** [review]
Did some testing now, and fetching the images is a bit slow but I'm not sure how that could be changed.. I noticed that if you try to save your profile image eventhough you haven't selected an image, it causes an error (cannot read properties of null)

**Comment by Jezerig** [comment]
- Crashes if you click 'save' when trying to add a profile picture but don't actually select any
- Crashes if you edit and save an event and don't add an image
- Crashes if you add too large profile picture (I tried 16mb as that is the max document size in MongoDB). We should probably limit to 4mb is something
- Crashes if you add too large  event image when editing
- Doesn't crash when adding too big image to a new event, but also doesn't add the image nor inform the user in any way

**Comment by Jezerig** [comment]
image

**Comment by jannejjj** [comment]
Sounds like we need a check for the image size, a check that an image has avtually been uploaded, and then remove the duplicate Event declaration from api.js :D

**Comment by EduardoLopezPalomo** [comment]
> * Crashes if you click 'save' when trying to add a profile picture but don't actually select any
> * Crashes if you edit and save an event and don't add an image
> * Crashes if you add too large profile picture (I tried 16MB as that is the max document size in MongoDB). We should probably limit it to something very small because for example a 4MB image took 500ms to load (GET /api/getImage/6611c8401196fdf4b1cc9a1b 304 427.412 ms - -)
> * Crashes if you add too large  event image when editing
> * Doesn't crash when adding too big image to a new event, but also doesn't add the image nor inform the user in any way

I solve all the bugs, but the last one the toast message is not shown because the page reloads immediately and i dont know how to solve it

**Comment by Jezerig** [comment]
image

**Comment by Jezerig** [comment]
- When saving profile image and an error occurs, the modal should not close (if possible)
- There should be a text informing the user what the max size of the image can be

**Comment by Jezerig** [comment]
- I don't think the event should be created if the image is too big, or if we want to do it like that, then the modal should also be closed
- The event is also created even if the if the file is not an image, and this should be communicated to the user as currently there is only an error message (if we want to do it this way)
- Choose wrong file type -> click cancel -> open modal again -> click save without choosing a new file -> gives the error message for the old file

**Comment by Jezerig** [comment]
image
Crashes if no file is chosen in profile picture, new event and when editing event.

File: client/src/components/EventItem.js [hunk 3]
```diff
@@ -184,6 +186,39 @@ const savingRules = () => {
...
+      const uint8Array = new Uint8Array(imageData.buffer.data);
+
+      // Convert the Uint8Array to a Base64 string
+      const base64String = uint8Array.reduce((data, byte) => data + String.fromCharCode(byte), '');
+      const imageUrl = `data:${imageData.mimetype};base64,${btoa(base64String)}`; 
+      setImage(imageUrl); 
+    }
+    else{
+      if(response.status == 413){
+        toasts.showToastMessage('The image size is to big');
```
Comment (Jezerig) [inline client/src/components/EventItem.js:217]: should be 'too big'

File: client/src/components/Home.js [hunk 4]
```diff
@@ -221,14 +227,52 @@ function Home(props) {
...
+              return;
+            }
+            const formData = new FormData();
+            formData.append('image', selectedFile);
+            const response = await fetch(`/api/updateImage/${eventId}`, {
+              method: 'POST',
+              body: formData
+            });
+            if(response.status == 413){
+              toasts.showToastMessage("the event is created but the image is to big");
```
Comment (Jezerig) [inline client/src/components/Home.js:259]: same here

File: client/src/components/MyProfile.js [hunk 5]
```diff
@@ -304,6 +309,64 @@ function ProfileItem(props) {
...
+        const imageUrl = `data:${imageData.mimetype};base64,${btoa(base64String)}`; 
+        setProfilePicture(imageUrl);
+        setUploadImage(false); 
+      }
+      else{
+        if(response.status == 409){
+          toasts.showToastMessage('Error while uploading the image');
+        }
+        if(response.status == 413){
+          toasts.showToastMessage('The image size is to big');
```
Comment (Jezerig) [inline client/src/components/MyProfile.js:352]: and here

File: server/routes/api.js [hunk 6]
```diff
@@ -669,6 +672,67 @@ router.get('/event/participants/:id',passport.authenticate('jwt', {session: fals
...
+      const event = await Event.findById(eventId);
+      if(!event){
+        return res.status(404).json({message: "Event not found"});
+      }
+      if(!file){
+        return res.status(409).json({error: "there is no image"})
+      }
+      const fileSize= file.size / (1024 * 1024); //size in megabytes
+      if(fileSize > maxSize){
+        return res.status(413).json({error: "the image size is to big"});
```
Comment (Jezerig) [inline server/routes/api.js:712]: here

File: server/routes/users.js [hunk 7]
```diff
@@ -30,6 +36,66 @@ router.get('/getData/:userId', async (req, res) => {
...
+    const maxSize = 2;
+    if(!user){
+      return res.status(404).json({message: "User not found"});
+    }
+    if(!file){
+      return res.status(409).json({error: "there is no image"})
+    }
+    const fileSize= file.size / (1024 * 1024); //size in megabytes
+    if(fileSize > maxSize){
+      return res.status(413).json({error: "the image size is to big"});
```
Comment (Jezerig) [inline server/routes/users.js:76]: here

File: client/src/components/Home.js (hunk 4 above)
Comment (Jezerig) [reply client/src/components/Home.js:259]: The first letter of a sentence should also be capitalized and all of these cases

Comment (Jezerig) [reply client/src/components/Home.js:259]: Also 'was created'

File: client/src/components/Home.js [hunk 8]
```diff
@@ -221,14 +227,52 @@ function Home(props) {
...
+              return;
+            }  
+            if (selectedFile.type !== 'image/png' && selectedFile.type !== 'image/jpeg') {
+              // Empty the input fields
+              setNewEvent({});
+              // Close the Modal
+              setNewEventModal(false);
+              // Update events list by toggling the boolean
+              toggleUpdateEvents();
+              toasts.showToastMessage('The event is created without an image');
```
Comment (Jezerig) [inline client/src/components/Home.js:249]: In the toast, inform the user that the file format was incorrect

**Comment by Jezerig** [comment]
> > * I don't think the event should be created if the image is too big, or if we want to do it like that, then the modal should also be closed
> 
> The problem is that i need the event id for connecting the image to the event. Another option is not allow to upload the image when creating a new event.

Ok, then it's fine how it is just with the toasts informing the user what happened

**Comment by Jezerig** [comment]
In the Scheduler, whichever photo is loaded first is shown for all the events. For example if I open event A from the calendar, then the photo of event A is shown for all other events.

**Comment by Enlarem** [comment]
> In the Scheduler, whichever photo is loaded first is shown for all the events. For example if I open event A from the calendar, then the photo of event A is shown for all other events.

It's because something is lack (around line 110) in the useEffect triggered by props updates

**Comment by Jezerig** [comment]
Now it's not possible to edit any event details without adding a new image :D

**Review by Jezerig — APPROVED** [review]

**Comment by jannejjj** [comment]
Seems to work nicely, I noticed a small typo though:
image

**Review by Nesssu — APPROVED** [review]

**Pull request merged**